hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/zlib/BuiltInGzipCompressor.java

```diff
@@ -0,0 +1,260 @@
+            } else {
+                return numAvailBytes;
+            }
+        }
+
+        numAvailBytes += writeTrailer(b, off, len);
+
+        return numAvailBytes;
+    }
+
+    @Override
+    public long getBytesRead() {
+        return deflater.getTotalIn();
+    }
+
+    @Override
+    public long getBytesWritten() {
+        return headerBytesWritten + deflater.getTotalOut() + trailerBytesWritten;
+    }
+
+    @Override
+    public void end() { deflater.end(); }
+
+    @Override
+    public void finish() { deflater.finish(); }
+
+    @Override
+    public void reinit(Configuration conf) {
+        ZlibCompressor.CompressionLevel level = ZlibFactory.getCompressionLevel(conf);
+        ZlibCompressor.CompressionStrategy strategy = ZlibFactory.getCompressionStrategy(conf);
+
+        // 'true' (nowrap) => Deflater will handle raw deflate stream only
+        deflater = new Deflater(level.compressionLevel(), true);
+        deflater.setStrategy(strategy.compressionStrategy());
+        state = BuiltInGzipDecompressor.GzipStateLabel.HEADER_BASIC;
+        crc.reset();
+        userBufOff = userBufLen = 0;
+        headerBytesWritten = trailerBytesWritten = 0;
+        currentInputLen = 0;
+        headerOff = 0;
+    }
+
+    @Override
+    public void reset() {
+        deflater.reset();
+        state = BuiltInGzipDecompressor.GzipStateLabel.HEADER_BASIC;
+        crc.reset();
+        userBufOff = userBufLen = 0;
+        currentInputLen = 0;
+        headerOff = 0;
+    }
+
+    @Override
+    public void setDictionary(byte[] b, int off, int len) {
+        deflater.setDictionary(b, off, len);
+    }
+
+    @Override
+    public void setInput(byte[] b, int off, int len) {
+        if (b == null) {
+            throw new NullPointerException();
+        }
+        if (off < 0 || len < 0 || off > b.length - len) {
+            throw new ArrayIndexOutOfBoundsException();
+        }
+
+        userBuf = b;
+        userBufOff = off;
+        userBufLen = len;
+
+        deflater.reset();
+        crc.reset();
+    }
+
+    private int writeHeader(byte[] b, int off, int len) {
+        if (len <= 0) {
+            return 0;
+        }
+
+        int n = Math.min(len, 10 - headerOff);
+        System.arraycopy(GZIP_HEADER, headerOff, b, off, n);
+        headerOff += n;
+
+        return n;
+    }
+
+    private int writeTrailer(byte[] b, int off, int len) {
+        int writtenSize = 0;
+
+        if (len >= 4 && state == BuiltInGzipDecompressor.GzipStateLabel.TRAILER_CRC) {
+            int streamCrc = (int) crc.getValue();
+            b[off] = (byte) (streamCrc & 0x000000ff);
+            b[off + 1] = (byte) ((streamCrc & 0x0000ff00) >> 8);
+            b[off + 2] = (byte) ((streamCrc & 0x00ff0000) >> 16);
+            b[off + 3] = (byte) ((streamCrc & 0xff000000) >> 24);
+
+            len -= 4;
+            off += 4;
+
+            writtenSize += 4;
+            state = BuiltInGzipDecompressor.GzipStateLabel.TRAILER_SIZE;
```

Comment:
what if we run out of buffer place and exit here with `BuiltInGzipDecompressor.GzipStateLabel.TRAILER_SIZE` state? will we enter into `compress` method again and write a header?

Reply:
Because I reserve buffer size for `len - GZIP_TRAILER_LEN`, I think we can expect in `writeTrailer` there is at least 8 bytes for writing the trailer.

Reply:
This was updated now. The trailer is written incrementally like the header.